src/io.ml

```diff
@@ -2,11 +2,94 @@ open Import
 
 module P = Pervasives
 
+module W32_io = struct
+
+  (* Pervasives.open* support neither O_CLOEXEC nor O_SHARE_DELETE *)
+
+  let trans accu = function
+  |	Open_rdonly -> Unix.O_RDONLY::accu
+  |	Open_wronly -> Unix.O_WRONLY::accu
+  |	Open_creat -> Unix.O_CREAT::accu
+  |	Open_trunc -> Unix.O_TRUNC::accu
+  |	Open_excl -> Unix.O_EXCL::accu
+  |	Open_nonblock -> Unix.O_NONBLOCK::accu
+  |	Open_append
+  |	Open_binary
+  |	Open_text -> accu
+
+  let sys_exn e fln =
+    let msg = fln ^ ":" ^ Unix.error_message e in
+    raise (Sys_error msg)
+
+  let eclose ecode fd fln =
+    (try Unix.close fd with Unix.Unix_error _ -> ());
+    sys_exn ecode fln
+
+  let open_gen mode perm fln =
+    let init = [Unix.O_SHARE_DELETE ; Unix.O_CLOEXEC] in
+    let m = List.fold_left ~f:trans ~init mode in
+    match Unix.openfile fln m perm with
+    | exception (Unix.Unix_error(x,_,_)) -> sys_exn x fln
+    | fd ->
+      if List.mem Open_append ~set:mode then (
+        try
+          let _ : int = Unix.lseek fd 0 Unix.SEEK_END in
+          ()
+        with
+        | Unix.Unix_error(e,_,_) -> eclose e fd fln
+      );
+      fd
+
+  let open_out_gen mode perm fln =
+    let fd = open_gen mode perm fln in
+    match Unix.out_channel_of_descr fd with
+    | exception (Unix.Unix_error(e,_,_)) -> eclose e fd fln
+    | oc ->
+      if List.mem Open_text ~set:mode then (
+        try
+          set_binary_mode_out oc false
+        with
+        | x -> close_out oc ; raise x
```

Comment:
I'm not totally convinced of the need to handle errors in `set_binary_mode_*`, but it should be `Unix.close fd` if you do